File: 3310.java
class Solution {
    public void dfs(int node, ArrayList<ArrayList<Integer>> graph,boolean [] vis)
    {
        vis[node]=true;
        for(int neigh:graph.get(node))
        {
            if(!vis[neigh])
            {
                dfs(neigh,graph,vis);
            }
        }

    }
    public List<Integer> remainingMethods(int n, int k, int[][] invocations) {
        
        ArrayList<ArrayList<Integer>> graph=new ArrayList<>();

        for(int i=0;i<n;i++)
        {
            graph.add(new ArrayList<>());
        }

        for(int [] edge: invocations)
        graph.get(edge[0]).add(edge[1]);

        boolean [] vis=new boolean[n];

        dfs(k,graph,vis);
        
        for(int [] edges:invocations)
        {
            int u=edges[0];
            int v=edges[1];
            if(!vis[u] && vis[v])
            {
                List<Integer> ans = new ArrayList<>();
                for(int i = 0; i < n; i++)
                    ans.add(i);
                return ans;
            }
        }

        List<Integer> ans=new ArrayList<>();
        for(int i=0;i<n;i++)
        {
            if(!vis[i])
            ans.add(i);
        }
    return ans;
    }
}

File: 3310.py
class Solution:
    def remainingMethods(self, n: int, k: int, invocations: List[List[int]]) -> List[int]:
        graph = defaultdict(list)

        for src, dest in invocations:
            graph[src].append(dest)

        visited = set()
        def dfs(node):
            visited.add(node)
            for child in graph[node]:
                if child not in visited:
                    dfs(child)

        dfs(k)
        all_methods = [i for i in range(n)]
        ret = []
        for v1, v2 in invocations:
            if v1 in visited and v2 not in visited or v2 in visited and v1 not in visited:
                return all_methods
        print(visited)
        return list(set(all_methods) - visited)